source/dagger/core/graphics/tool_render.cpp

```diff
@@ -0,0 +1,163 @@
+#include "tool_render.h"
+#include "sprite.h"
+#include "textures.h"
+#include "texture.h"
+#include "core/engine.h"
+
+#include <limits>
+#include <algorithm>
+#include <execution>
+#include <chrono>
+#include <cstdint>
+#include <regex>
+
+using namespace dagger;
+
+void ToolRenderSystem::SpinUp()
+{
+    glGenVertexArrays(1, &m_VAO);
+    glBindVertexArray(m_VAO);
+
+    glGenBuffers(1, &m_StaticMeshVBO);
+    glBindBuffer(GL_ARRAY_BUFFER, m_StaticMeshVBO);
+    glBufferData(GL_ARRAY_BUFFER, sizeof(m_VerticesAndTexCoords), m_VerticesAndTexCoords, GL_STATIC_DRAW);
+
+    // attribute #0: vertex position
+    glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE,
+        sizeof(float) * 4, (void*)0);
+    glEnableVertexAttribArray(0);
+
+    // attribute #1: tex coord
+    glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE,
+        sizeof(float) * 4, (void*)(sizeof(float) * 2));
+    glEnableVertexAttribArray(1);
+
+    const GLsizei dataSize = sizeof(SpriteData);
+
+    glGenBuffers(1, &m_InstanceQuadInfoVBO);
+    glBindBuffer(GL_ARRAY_BUFFER, m_InstanceQuadInfoVBO);
+    glBufferData(GL_ARRAY_BUFFER, s_BufferSize, nullptr, GL_STREAM_DRAW);
+
+    const StaticArray<Pair<UInt32, UInt32>, 9> sizesAndStrides = {
+        pair(2, 0),     // #2: sub size
+        pair(2, 2),     // #3: sub origin
+        pair(2, 4),     // #4: sub origin
+        pair(3, 6),     // #5: quad position
+        pair(2, 9),     // #6: quad pivot
+        pair(4, 11),    // #7: quad tint color
+        pair(2, 15),    // #8: scale
+        pair(1, 17),    // #9: rotation
+        pair(1, 18),    // #10: is UI?
+    };
+
+    for (UInt32 i = 0; i < sizesAndStrides.size(); i++)
+    {
+        const UInt32 index = 2 + i;
+        glVertexAttribPointer(index, sizesAndStrides[i].first, GL_FLOAT, GL_FALSE,
+            dataSize, (void*)(sizeof(Float32) * sizesAndStrides[i].second));
+        glEnableVertexAttribArray(index);
+        glVertexAttribDivisor(index, 1);
+    }
+
+    glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+    glEnable(GL_TEXTURE_2D);
+    glActiveTexture(GL_TEXTURE0);
+
+    Engine::Dispatcher().sink<Render>().connect<&ToolRenderSystem::OnRender>(this);
+}
+
+void ToolRenderSystem::OnRender()
+{
+    static auto SortSprites = [](const Sprite& a_, const Sprite& b_)
+    {
+        UInt32 aShader = a_.shader->programId;
+        UInt32 bShader = b_.shader->programId;
+        UInt32 aZ = a_.position.z;
+        UInt32 bZ = b_.position.z;
+        UInt32 aImage = a_.image == nullptr ? 0 : a_.image->TextureId();
+        UInt32 bImage = b_.image == nullptr ? 0 : b_.image->TextureId();
+
+        if (aZ == bZ)
+        {
+            if (aShader == bShader)
+            {
+                return aImage < bImage;
+            }
+            else
+                return aShader < bShader;
+        }
+        else
+            return aZ > bZ;
```

Comment:
I'm not sure where are the cases of aZ>bZ and aZ<bZ covered here. I suppose they are indirectly through the other variables or they're irrelevant or something, it would just be nice if there was a clearer if-else or comments explaining it.

Reply:
There is no case where they're covered: these are different levels of sorting -- we have z-order, shader and texture as the three parameters of our sort. If two elements are the same, we look at the next level. If they aren't, the ordering of the current level is used. :) this is the same code from sprite renderer, but I'll document it anyway!

Reply:
Ahh I get it now, it wasn't especially complex actually. I think I just got confused because I have a habit of checking inequalities first and then doing the equality in an else branch, boils down to the same thing and it's shorter to write like this. Thanks for documenting anyway!